Fundamentals/Setup/Server-Setup/runtime-modes.md

```diff
@@ -0,0 +1,105 @@
+- The application is built/published in Release mode (with JIT optimization enabled), e.g. using `dotnet publish --configuration Release`;
+- `Umbraco:CMS:WebRouting:UmbracoApplicationUrl` is set to a valid URL;
+- `Umbraco:CMS:Global:UseHttps` is enabled;
+- `Umbraco:CMS:RuntimeMinification:CacheBuster` is set to a fixed cache buster like `Version` or `AppDomain`;
+- `Umbraco:CMS:ModelsBuilder:ModelsMode` is set to `Nothing`.
+
+The recommended approach to enable `Production` mode is to update the `appsettings.Production.json` file (or create one) with the following settings:
+
+```json
+{
+  "Umbraco": {
+    "CMS": {
+      "Runtime": {
+        "Mode": "Production"
+      },
+      "Global": {
+        "UseHttps": true
+      },
+      "ModelsBuilder": {
+        "ModelsMode": "Nothing"
+      },
+      "WebRouting": {
+        "UmbracoApplicationUrl": "https://<REPLACE_WITH_YOUR_PRIMARY_DOMAIN>/"
+      }
+    }
+  }
+}
+```
+
+
+Although you can still edit document types and views (if not running from the published output), changes won't be picked up until you've rebuilt your project or republished the application. Also note that models won't be generated by ModelsBuilder (because the mode is set to `Nothing`), requiring your to do all your changes while in `Development` mode.
+
+Also ensure the `UmbracoApplicationUrl` is updated to the primary URL of your production environment, as this is used when sending emails (password reset, notifications, health check results, etc.) and the keep-alive task.
+
+## Customize/extend runtime mode validation
+
+Validation of the above mentioned settings are done when determining the runtime level during startup using the new `IRuntimeModeValidationService` and when it fails, causes a `BootFailedException` to be thrown. The default implementation gets all registered `IRuntimeModeValidators` to do the validation, making it possible to remove default checks and/or add your own (inherit from `RuntimeModeProductionValidatorBase` if you only want to validate against the production runtime mode). The following validators are added by default:
+- `JITOptimizerValidator` - Ensure the application is built/published in Release mode (with JIT optimization enabled) when in production runtime mode, e.g. using `dotnet publish --configuration Release`;
+- `UmbracoApplicationUrlValidator` - ensure `Umbraco:CMS:WebRouting:UmbracoApplicationUrl` is configured when in production runtime mode;
+- `UseHttpsValidator` - ensure `Umbraco:CMS:Global:UseHttps` is enabled when in production runtime mode;
+- `RuntimeMinificationValidator` - ensure `Umbraco:CMS:RuntimeMinification:CacheBuster` is set to a fixed cache buster like `Version` or `AppDomain` when in production runtime mode;
+- `ModelsBuilderModeValidator` - ensure `Umbraco:CMS:ModelsBuilder:ModelsMode` is not set to `InMemoryAuto` when in development runtime mode and set to `Nothing` when in production runtime mode.
```

Comment:
Now that PR https://github.com/umbraco/Umbraco-CMS/pull/12693 is merged, we can add the following sample code to show how you can remove default and/or add custom runtime mode validators:

```suggestion
- `ModelsBuilderModeValidator` - ensure `Umbraco:CMS:ModelsBuilder:ModelsMode` is not set to `InMemoryAuto` when in development runtime mode and set to `Nothing` when in production runtime mode.

The following example removes the default `UmbracoApplicationUrlValidator` and adds a new custom `DisableElectionForSingleServerValidator`:

    using System.Diagnostics.CodeAnalysis;
    using Microsoft.Extensions.Options;
    using Umbraco.Cms.Core.Composing;
    using Umbraco.Cms.Core.Configuration.Models;
    using Umbraco.Cms.Infrastructure.Runtime;
    using Umbraco.Cms.Infrastructure.Runtime.RuntimeModeValidators;
    
    public class RuntimeModeValidatorComposer : IComposer
    {
        public void Compose(IUmbracoBuilder builder)
            => builder.RuntimeModeValidators()
                .Remove<UmbracoApplicationUrlValidator>()
                .Add<DisableElectionForSingleServerValidator>();
    }
    
    public class DisableElectionForSingleServerValidator : IRuntimeModeValidator
    {
        private readonly IOptionsMonitor<GlobalSettings> _globalSettings;
    
        public DisableElectionForSingleServerValidator(IOptionsMonitor<GlobalSettings> globalSettings) => _globalSettings = globalSettings;
    
        public bool Validate(RuntimeMode runtimeMode, [NotNullWhen(false)] out string? validationErrorMessage)
        {
            if (_globalSettings.CurrentValue.DisableElectionForSingleServer == false)
            {
                validationErrorMessage = "Disable primary server election (and support for load balancing) to improve startup performance.";
                return false;
            }
    
            validationErrorMessage = null;
            return true;
        }
    }
```

Reply:
I've indented the sample code 4 spaces (as mentioned in the Markdown conventions for our Docs), but other code blocks still use ` ```json` and ` ```xml` style markup...

Reply:
@ronaldbarendse, you can either use 4 spaces or add an optional language identifier to enable syntax highlighting in your fenced code block. We might need to update the Markdown conventions article with this info 🙂